intel_adsp: ace: dynamic clock switching

tmleman opened the pull request:
This pull request introduces a series of enhancements to the power management capabilities of the Intel ADSP platforms. The changes include the implementation of dynamic clock switching, which allows the DSP to adjust its clock frequency based on the idle state of the cores, thereby reducing power consumption during idle periods.

Depends on:

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/64468
- [ ] https://github.com/zephyrproject-rtos/zephyr/pull/66295

Here you can check SOF part of this feature: https://github.com/thesofproject/sof/pull/8423

tmleman committed:
soc: intel_adsp: implement dynamic clock switching

This patch introduces dynamic clock switching for Intel ADSP platforms,
enabling the DSP to adjust its clock frequency dynamically based on core
activity. The mechanism is designed to reduce power consumption by
lowering the clock frequency when cores enter an idle state and restoring
it when they become active again.

Key changes include:
- Incrementing a clock switching lock for each active core to prevent
  clock frequency changes during core activity.
- adsp_clock_idle_entry: Invoked when the DSP is about to enter an
  idle state, reducing the clock frequency to the minimum if not already
  there.
- adsp_clock_idle_exit: Called upon idle exit to restore the clock to the
  previous frequency.
- Utilizing a spinlock to protect clock frequency changes from concurrent
  access.

This feature is enabled with the CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
configuration option.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

Commit:
pm: state: Support setting custom data to a state

Custom data is not used by the subsystem itself, and is freely available
for a SoC to use as it sees fit. It can be used to SoCs expose a way
to targets customize a state behavior.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
(cherry picked from commit ae69368a6481ad4ec402bf4537bfd11ff3aec63e)

tmleman commented:
I'm changing it temporarily to draft.

After changes in https://github.com/zephyrproject-rtos/zephyr/pull/64468 I need to rework this PR. The new implementation will likely be based on https://github.com/zephyrproject-rtos/zephyr/pull/66295.

tmleman committed:
soc: intel_adsp: ace: add PM data structure for clock management

Introduce a new header file, pm_data.h, for the Intel ADSP ACE platforms,
which defines a structure and inline functions for managing power
management-related data, specifically for clock switching operations.

The structure intel_adsp_ace_pm_data includes an atomic reference
counter to control clock switching lock states, ensuring thread-safe
operations across multiple cores.

Two inline functions, intel_adsp_acquire_clock_switch_lock and
intel_adsp_release_clock_switch_lock, are provided to safely modify the
lock reference counter, with assertions to prevent unbalanced lock usage.

The header file for CAVS currently serves only as a placeholder. This is
intended to avoid a situation where including this header would need to
be conditional with an ifdef.

This addition is a foundational step for implementing more sophisticated
power management capabilities for the Intel ADSP ACE platforms.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
soc: intel_adsp: add dynamic clock switching support

This patch introduces the ADSP_DYNAMIC_CLOCK_SWITCHING Kconfig option,
enabling firmware to dynamically adjust the DSP clock frequency when all
cores are idle. Aimed at reducing energy consumption, this feature
minimizes power usage during idle periods without significantly
increasing resume latency.

The option is configurable via Kconfig. When enabled, it complements the
existing ADSP_IDLE_CLOCK_GATING feature by providing additional power
savings while maintaining quick resume times from idle states.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
@ceolin please check the latest updates.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -381,4 +381,28 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+
+#include <pm_data.h>
+
+void arch_cpu_idle(void)
```
Instead of having it, couldn't you use a new power state  ? A new runtime-idle with a different substate  and residency time ? Then you just need to add it in the existent `pm_state_set`

tmleman replied:
At first, I thought about it and considered it a stylistically nicer solution. However, in this case, all I have to do is change the clock and then transition to the WAITI state. I'm concerned that the entire logic provided by the execution of code from `pm_system_suspend` may result in us losing any potential benefits that this entire operation is supposed to deliver.

ceolin commented on soc/xtensa/intel_adsp/common/clk.c:
```diff
@@ -153,3 +153,57 @@ uint32_t adsp_clock_source_frequency(int source)
+{
+	/* clock has not been changed */
+	if (!atomic_get(&clock_change_count))
+		return;
```
This atomic is not really protecting this area, but it does not even seen to be necessary, the calls for these functions are already synced.

ceolin commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -381,4 +381,28 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+	struct intel_adsp_ace_pm_data *data =
+		(struct intel_adsp_ace_pm_data *) pm_state_custom_data_get(PM_STATE_ACTIVE, 0);
+
+	if (data && decrement_clock_switch_lock(data))
```
This check is not enough to protect this region .. another core can successfully ` increment_clock_switch_lock(data)` and call `adsp_clock_idle_exit();` after this check and before this core executes `adsp_clock_idle_entry();` You need probably a spin lock around these operations.

andyross replied:
Agreed, this looks racy.

tmleman replied:
It seems I didn't think it through carefully enough. Thanks.

andyross commented on include/zephyr/pm/state.h:
```diff
@@ -154,6 +154,14 @@ struct pm_state_info {
+	 * Target specific custom data. Support for runtime customization of
+	 * a particular idle state.
+	 */
+	const void *custom_data;
```
Not a fan of the "stuff a void* into a struct" paradigm; untyped APIs are dangerous.   In situations where we have a need for platform-specific additional data in core kernel sturcts, zephyr typically adds a substruct, e.g. something like:

```
IF_ENABLED(CONFIG_SOC_HAS_PM_STATE, struct soc_pm_state soc);
```

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -381,4 +381,13 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+void arch_cpu_idle(void)
+{
+	sys_trace_idle();
+	__asm__ volatile("waiti 0");
```
Just remove this patch?  Just add your custom idle code once, no need to add a noop copy first.

andyross replied:
Also, and I do see that the ARCH_CPU_IDLE_CUSTOM has already merged, but the whole feature looks silly to me.  The Zephyr kernel **already** allows for indirection of CPU idle entry to the platform layers, that's why it's called "arch_"!

If the arch wants to further delegate, that's fine too, but there's no reason the core Zephyr code should need to care.

In this case, it would have been much simpler to make the xtensa arch_cpu_idle() a `__weak` function, document that fact, and then just implement your own in the SOC code.

andyross commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -383,10 +383,25 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 {
+#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
+	extern void adsp_clock_idle_entry(void);
+	extern void adsp_clock_idle_exit(void);
```
An occasional extern declaration for a weakly-coupled internal function is sort of OK as a way to punch a hole into a separate subsystem.  But this is a real API with clear use and obvious applicability to the code here.  These declarations should be in a header under soc.h somewhere.

andyross commented on soc/xtensa/intel_adsp/common/boot.c:
```diff
@@ -16,6 +16,11 @@
 #include <cpu_init.h>
 #include "manifest.h"
 
+#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
```
Nitpick: #if'ing headers is usually pointless.  C headers are cheap to parse, and with very few exceptions all Zephyr headers should be includable in any context regardless of build-time configuration.

marc-hb replied:
.h files is one of C/C++ worst headaches:
- https://github.com/zephyrproject-rtos/zephyr/issues/41543

>  all Zephyr headers should be includable in any context regardless of build-time configuration.

Indeed, one way to help maintain this property is to parse headers even when they're not in use because this detects conflicts and other problems SOONER, not much later when someone turns on some rarely use CONFIG_

tmleman replied:
This feature is not enabled on CAVS, and this header needs to be under #ifdef in this common file.

marc-hb replied:
I just tried and solved this problem by adding an empty `soc/xtensa/intel_adsp/cavs/include/pm_data.h` placeholder.

Simpler, less unusual and more consistent. Now `ace/include/pm_data.h` is parsed every time which provides more coverage and can find more conflicts.

You just need to add at least one line to pass CI checks: good opportunity to mention the purpose of the placeholder.

tmleman replied:
@marc-hb thanks for the hint.

andyross reviewed (commented):
I'd actually been trying to avoid doing review on Intel hardware stuff, but now I've been sucked in to patch juggling on the SOF MTL release branch (which has an ancestor of this work) and I don't have a choice :(

Some notes and nitpicks.  Really the big one is this: why is this a platform API at all?  The core feature is a refcount on PM state that indicates the number of "users" that need the core alive.  It's cheaply implemented with just one atomic word and two inline functions.

Can't we put this in the core PM layer?  Almost everyone is going to want this, and those who don't won't pay much cost even if we don't make it kconfig-guarded.

andyross commented on soc/xtensa/intel_adsp/ace/include/pm_data.h:
```diff
@@ -0,0 +1,73 @@
+ * @param data Pointer to the intel_adsp_ace_pm_data struct containing the lock reference counter.
+ * @return True if this is the first increment operation, false otherwise.
+ */
+static ALWAYS_INLINE bool increment_clock_switch_lock(struct intel_adsp_ace_pm_data *data)
```
Another nitpick: these two functions want to be namespaced to intel_adsp_ (or equivalent), the SOC headers tend to be included very widely even when you don't want them to. But again I really think we want this in the core PM layer instead.

tmleman commented:
@ceolin @andyross Please take another look.

marc-hb commented on soc/xtensa/intel_adsp/ace/multiprocessing.c:
```diff
@@ -20,6 +20,11 @@
+#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
+#include <zephyr/pm/policy.h>
+#include <pm_data.h>
+#endif
```
Drop the conditional for more coverage, see below. This gives Zephyr CI and others a chance to find issues before other CIs do.

BTW don't forget to submit this to SOF CI with a one-line `sof/west.yml` change to catch problems earlier, unlike what happened in

- https://github.com/zephyrproject-rtos/zephyr/pull/67393#issuecomment-1889640994

tmleman replied:
The pull request to SOF was already made earlier. The link is in the description.

marc-hb commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -397,4 +397,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 
+#ifdef CONFIG_ARCH_CPU_IDLE_CUSTOM
+
+#include <pm_data.h>
```
Include unconditionally at the top.

tmleman commented:
Currently, I cannot perform a rebase because I will lose the ability to build this branch along with SOF. Additionally, there is an issue when bringing the main core down to the D3 state (DSP panic during memcpy - debugging in progress).

tmleman commented:
Rebase

tmleman committed:
dts: xtensa: intel_adsp_ace15_mtpm: Add active power state to CPUs

Include the 'active' power state in the DTS for all CPU cores on the
Intel ADSP ACE15 (MTL) platform to enable setting custom power state
data. Without this change, the system is unable to set and retrieve
custom data associated with the active power state, as the state would
not be present in the cpus_states array used by the power management
subsystem.

The addition of the active state in the DTS is critical for systems that
utilize the CONFIG_PM_STATE_CUSTOM_DATA configuration, as it allows the
pm_state_custom_data_set and pm_state_custom_data_get functions to
associate and access custom data with the active state.

This update ensures that the active state, which is expected to be
always available for CPU cores, is correctly reflected in the power
management arrays, thus facilitating accurate power state management and
transitions.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
pm: Fix retrieval of custom power state data

Correct the implementation of pm_state_custom_data_get to properly
retrieve custom data associated with specified power states. The
previous version of the function always returned null.

The patch replaces the direct array reference with a call to
pm_state_cpu_get_all, which fetches the correct CPU power states based
on the current processor ID. This ensures that the function iterates
over the appropriate set of power states when searching for custom data.

It is currently a draft intended to enable testing and further
development.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
dts: xtensa: intel_adsp_ace20_lnl: Add active state to CPU power states

Include the 'active' power state in the DTS for all CPU cores on the
Intel ADSP ACE20 (LNL) platform, paralleling the update made for the MTL
platform. This addition is necessary to enable the setting and retrieval
of custom power state data associated with the active state, which would
otherwise be absent from the cpus_states array and unavailable to the
power management subsystem.

Defining the active state in the DTS is crucial for systems that rely on
the CONFIG_PM_STATE_CUSTOM_DATA configuration. It permits the
pm_state_custom_data_set and pm_state_custom_data_get functions to work
with custom data linked to the active state.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented on soc/xtensa/intel_adsp/Kconfig:
```diff
@@ -133,4 +133,12 @@ config ADSP_IDLE_CLOCK_GATING
+	bool "DSP clock switching in Idle"
+	depends on PM_STATE_CUSTOM_DATA
+	help
+	  When true, FW will switch clock if all cores are in idle state. The dynamic
```
@marc-hb Please take a look at the updated version to see if it's what you had in mind?

marc-hb replied:
That's a perfect Kconfig help thank you. I wish they were all like this :-)

Now I wonder why it's not enabled by default but I guess it's either because it's a bit early still or because I missed something else so nevermind.

In case you were wondering, `default y` does NOT override `depends on PM_STATE_CUSTOM_DATA`.

tmleman commented:
@ceolin @andyross @marc-hb I finally solved the problem in the `imr context save` case. The branch has been rebased, and I invite you to review.

ceolin commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -25,17 +25,23 @@
+				power-state-name = "active";
+				min-residency-us = <0>;
+				exit-latency-us = <0>;
+			};
```
I particularly don't think that is a good thing. Declare this state here is to workaround the problem.

I think one way to approach the problem without need the custom data is creating a proper state (runtime-idle with some substate, or any other state ...) with minimal residency. Then the application can put/release locks in this state if it wants to switch clock on idle.

Another possibility and clear (?) is adding a proper API in the adsp clock control (look other comments)

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -154,6 +154,14 @@ struct pm_state_info {
 	 */
 	uint32_t exit_latency_us;
+
+#ifdef CONFIG_PM_STATE_CUSTOM_DATA
```
Thinking more about it I believe it is a bad idea, at least for this case. The data is quite dependent of the target and the way it is being used in SoF ended up calling target specific API (` intel_adsp_acquire_clock_switch_lock`).
It does not make sense having this workaround if you need to call that API anyways.

ceolin commented on soc/intel/intel_adsp/ace/include/pm_data.h:
```diff
@@ -0,0 +1,73 @@
+
+	return false;
+}
+
```
Can't / Shouldn't we move this to https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/intel/intel_adsp/common/include/adsp_clk.h ?

And change names to something like

`adsp_clock_switch_on_idle_enable` 
`adsp_clock_switch_on_idle_disable`
`adsp_clock_switch_on_idle_is_enable`

This header is already exported through https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/clock_control/clock_control_adsp.h

ceolin commented on soc/intel/intel_adsp/ace/power.c:
```diff
@@ -439,11 +440,25 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+		(struct intel_adsp_ace_pm_data *) pm_state_custom_data_get(PM_STATE_ACTIVE, 0);
+
+	k = k_spin_lock(&idle_lock);
+	if (data && intel_adsp_release_clock_switch_lock(data))
```
It is missing braces

ceolin commented on soc/intel/intel_adsp/ace/power.c:
```diff
@@ -456,6 +471,13 @@ void arch_cpu_idle(void)
+		soc_adsp_clock_idle_exit();
+
+	k_spin_unlock(&idle_lock, k);
+#endif /* CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING */
```
And this block id probably not needed.

ceolin commented on soc/intel/intel_adsp/cavs/include/pm_data.h:
```diff
@@ -0,0 +1,15 @@
+ * where including this header would need to be conditional with an ifdef.
+ */
+
+#endif /* __INTEL_PM_DATA_H__ */
```
no needed for this header.

marc-hb replied:
@ceolin this placeholder avoids an `#idef ... #include` in multiprocessing.c (or anywhere else). This was discussed in https://github.com/zephyrproject-rtos/zephyr/pull/66092#discussion_r1460105808 and others.

```C
#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
#include <pm_data.h>
#endif
```

ceolin replied:
sorry, I was not clear about it. I don't think the whole idea of pm_data is needed, so this header is shouldn't exist.

ceolin commented on soc/intel/intel_adsp/common/clk.c:
```diff
@@ -153,3 +153,32 @@ uint32_t adsp_clock_source_frequency(int source)
+		select_cpu_clock_hw(platform_cpu_clocks[0].lowest_freq);
+	}
+
+	k_spin_unlock(&lock, k);
```
aren't you already protecting this call with `idle_lock` ?

You see ... this function is already implemented in `clk.c` the functions enabling/disabling this behavior could be directly exported in adsp_clk header.

ceolin commented on soc/intel/intel_adsp/common/clk.c:
```diff
@@ -153,3 +153,32 @@ uint32_t adsp_clock_source_frequency(int source)
+		select_cpu_clock_hw(platform_cpu_clocks[0].current_freq);
+	}
+
+	k_spin_unlock(&lock, k);
```
same regarding the lock.

ceolin commented on subsys/pm/state.c:
```diff
@@ -66,3 +68,40 @@ uint8_t pm_state_cpu_get_all(uint8_t cpu, const struct pm_state_info **states)
+#endif
+
+	return NULL;
+}
```
This whole custom data is not needed doing that way.

ceolin reviewed (changes requested):
@tmleman sorry for the late reply. Now that I saw the workflow (including SoF side), I think there is a simpler way to do this. Please look the comments and check if they make sense for you.

andyross commented on soc/intel/intel_adsp/ace/power.c:
```diff
@@ -456,6 +471,13 @@ void arch_cpu_idle(void)
 	__asm__ volatile ("waiti 0");
+#if CONFIG_ADSP_DYNAMIC_CLOCK_SWITCHING
+	k = k_spin_lock(&idle_lock);
+	if (data && intel_adsp_acquire_clock_switch_lock(data))
```
What are the circumstances where data might be NULL?  This seems like a needless check to me that could be better expressed as an assert?

Also the MISRA scanner is snooping around again and it doesn't like the "implicitly convert pointer to boolean" idiom and demands this be expressed as "data == NULL".  But maybe it won't find its way into the soc code for a while.

andyross reviewed (commented):
Some drive-by notes.

tmleman commented:
@ceolin I started implementing the fixes suggested in the review and encountered two problems.

The first is tracking whether all cores are in the Idle state. To change the clock in the substate rt_idle, I would need to pull the prevent for this substate in the custom implementation of the Idle function (this also involves applying the prevent and restoring the clock when any of the cores exit the Idle state). This would look similar to what was in the first version of this PR.

The second problem is the time needed to change the clock while in Idle. It seems to me that doing this through pm_system_suspend introduces so much overhead logic that the small energy savings this feature provides would drop to zero or even have a negative effect. I don't have any numbers to support this thesis, so if you want, you can ignore it, but please address the first problem because I don't want to introduce unnecessary changes.

ceolin commented:
> @ceolin I started implementing the fixes suggested in the review and encountered two problems.
> 
> The first is tracking whether all cores are in the Idle state. To change the clock in the substate rt_idle, I would need to pull the prevent for this substate in the custom implementation of the Idle function (this also involves applying the prevent and restoring the clock when any of the cores exit the Idle state). This would look similar to what was in the first version of this PR.

I may have being confuse in my review. What I have proposed is to simplify it removing the custom data and the active power states that are not making sense, returning ACTIVE is the same as the policy returning false (no power state meets the time requirement).

> 
> The second problem is the time needed to change the clock while in Idle. It seems to me that doing this through pm_system_suspend introduces so much overhead logic that the small energy savings this feature provides would drop to zero or even have a negative effect. I don't have any numbers to support this thesis, so if you want, you can ignore it, but please address the first problem because I don't want to introduce unnecessary changes.

What I've suggest does not change when / where you are changing the clock.

Please take a look in this branch https://github.com/ceolin/zephyr/tree/temp-tomasz . That is what I had in mind.

tmleman commented:
Closing this. I might return to it later.
